File: src/services/rag/ragService.js
/**
 * RAG Orchestration Service
 * Ties together: embedding → vector search → reranking → LLM answer generation.
 */
const embeddingService = require('../embeddings/embeddingService');
const llmService = require('../llm/llmService');
const rerankService = require('./rerankService');
const vectorStoreService = require('../../vectorstore/vectorStoreService');
const cacheService = require('../../utils/cacheService');
const { parseDocument } = require('../../utils/documentParser');
const { chunkText, createChunkMetadata } = require('../../utils/chunkingService');
const { v4: uuidv4 } = require('uuid');

// ──────────────────────────────────────────
// Document Ingestion Pipeline
// ──────────────────────────────────────────

/**
 * Full upload pipeline: parse → chunk → embed → store.
 * @param {Object} file - Multer file object { originalname, buffer, mimetype }
 * @param {Object} opts - { chunkSize, chunkOverlap }
 * @returns {Promise<Object>} { documentId, filename, chunkCount, message }
 */
async function ingestDocument(file, opts = {}) {
  const documentId = uuidv4();
  const filename = file.originalname;
  const startTime = Date.now();

  console.log(`[RAG] Ingesting "${filename}" (id: ${documentId})`);

  // 1. Parse the document into raw text
  const rawText = await parseDocument(file);
  console.log(`[RAG]   Parsed: ${rawText.length} characters`);

  // 2. Chunk the text
  const chunks = chunkText(rawText, {
    chunkSize: opts.chunkSize,
    chunkOverlap: opts.chunkOverlap
  });
  console.log(`[RAG]   Chunked: ${chunks.length} chunks`);

  if (chunks.length === 0) {
    throw new Error('Document produced zero chunks after parsing');
  }

  // 3. Create chunk metadata
  const chunkMeta = createChunkMetadata(documentId, filename, chunks);

  // 4. Generate embeddings for all chunks (batched)
  const embeddings = await embeddingService.generateEmbeddingsBatched(
    chunks,
    20 // batch size
  );
  console.log(`[RAG]   Embedded: ${embeddings.length} vectors`);

  // 5. Store in ChromaDB
  await vectorStoreService.storeChunks(chunkMeta, embeddings);

  // 6. Invalidate query cache since document corpus changed
  cacheService.invalidateQueries();

  const elapsed = ((Date.now() - startTime) / 1000).toFixed(2);
  console.log(`[RAG]   Ingestion complete in ${elapsed}s`);

  return {
    documentId,
    filename,
    chunkCount: chunks.length,
    characterCount: rawText.length,
    processingTime: `${elapsed}s`,
    message: `Successfully ingested "${filename}" into ${chunks.length} chunks`
  };
}

// ──────────────────────────────────────────
// Query Pipeline
// ──────────────────────────────────────────

/**
 * Full query pipeline: embed query → search → rerank → generate answer.
 * @param {string} query - User question
 * @param {Object} opts - { topK, documentId, temperature, includeMetadata, verify, rerank, skipCache }
 * @returns {Promise<Object>} { answer, sources, query, processingTime, verification? }
 */
async function queryDocuments(query, opts = {}) {
  const topK = opts.topK || 5;
  const shouldVerify = opts.verify === true;
  const shouldRerank = opts.rerank === true;
  const skipCache = opts.skipCache === true || shouldVerify; // Don't cache verified queries
  const startTime = Date.now();

  // Check query cache first (only for non-verify requests)
  if (!skipCache) {
    const cacheOpts = { topK, documentId: opts.documentId, rerank: shouldRerank };
    const cached = cacheService.getQueryResult(query, cacheOpts);
    if (cached) {
      console.log(`[RAG] Query served from cache: "${query.substring(0, 50)}..."`);
      return {
        ...cached,
        processingTime: '0.00s (cached)'
      };
    }
  }

  console.log(`[RAG] Query: "${query.substring(0, 80)}..." (topK=${topK}, verify=${shouldVerify}, rerank=${shouldRerank})`);

  // 1. Generate embedding for the query
  const queryEmbedding = await embeddingService.generateEmbedding(query);

  // 2. Search for relevant chunks (fetch more if reranking)
  const fetchK = shouldRerank ? Math.min(topK * 3, 20) : topK;
  let results;
  if (opts.documentId) {
    results = await vectorStoreService.searchByDocument(queryEmbedding, opts.documentId, fetchK);
  } else {
    results = await vectorStoreService.searchSimilar(queryEmbedding, fetchK);
  }

  if (results.length === 0) {
    return {
      answer: 'No relevant documents found for your query. Please upload documents first or refine your question.',
      confidence: { score: 0, reason: 'No relevant documents found', level: 'none' },
      sources: [],
      query,
      reranked: false,
      processingTime: `${((Date.now() - startTime) / 1000).toFixed(2)}s`
    };
  }

  // 3. Optionally re-rank using cross-encoder
  if (shouldRerank && results.length > 1) {
    console.log(`[RAG]   Re-ranking ${results.length} chunks...`);
    results = await rerankService.rerankChunks(query, results, { topN: topK });
  } else if (results.length > topK) {
    // Trim to topK if we fetched extra but didn't rerank
    results = results.slice(0, topK);
  }

  // 4. Generate answer with LLM using the retrieved context
  const rawAnswer = await llmService.generateAnswer(query, results, {
    temperature: opts.temperature
  });

  // 5. Parse confidence from the response
  const { answer, confidence } = llmService.parseConfidence(rawAnswer);

  // 6. Optionally verify the answer against sources
  let verification = null;
  if (shouldVerify) {
    console.log(`[RAG]   Running answer verification...`);
    verification = await llmService.verifyAnswer(answer, results);
    console.log(`[RAG]   Verification complete: ${verification.isVerified ? 'PASSED' : 'FAILED'} (${verification.overallScore}/10)`);
  }

  // 7. Build source citations
  const sources = results.map(r => ({
    filename: r.metadata?.filename,
    chunkIndex: r.metadata?.chunkIndex,
    documentId: r.metadata?.documentId,
    similarityScore: parseFloat((r.similarityScore || 0).toFixed(4)),
    ...(r.crossEncoderScore !== undefined && {
      crossEncoderScore: r.crossEncoderScore,
      originalRank: r.originalRank
    }),
    ...(opts.includeMetadata !== false && {
      preview: r.text?.substring(0, 200) + (r.text?.length > 200 ? '...' : '')
    })
  }));

  const elapsed = ((Date.now() - startTime) / 1000).toFixed(2);
  console.log(`[RAG]   Query answered in ${elapsed}s using ${results.length} chunks (confidence: ${confidence.score}/10)`);

  const response = {
    answer,
    confidence,
    sources,
    query,
    topK,
    reranked: shouldRerank,
    chunksUsed: results.length,
    processingTime: `${elapsed}s`
  };

  // Include verification results if requested
  if (verification) {
    response.verification = verification;
  }

  // Cache the result (only if not skipping cache)
  if (!skipCache) {
    const cacheOpts = { topK, documentId: opts.documentId, rerank: shouldRerank };
    cacheService.setQueryResult(query, cacheOpts, response);
  }

  return response;
}

// ──────────────────────────────────────────
// Compare Pipeline
// ──────────────────────────────────────────

/**
 * Compare two documents on a given topic.
 * @param {string[]} documentIds - Exactly 2 document IDs
 * @param {string} topic - What to compare
 * @param {Object} opts - { topK, structured }
 * @returns {Promise<Object>} { comparison, doc1Sources, doc2Sources, topic, processingTime }
 */
async function compareDocuments(documentIds, topic, opts = {}) {
  const topK = opts.topK || 5;
  const useStructured = opts.structured === true;
  const startTime = Date.now();

  console.log(`[RAG] Comparing docs [${documentIds.join(', ')}] on topic: "${topic.substring(0, 60)}" (structured=${useStructured})`);

  // 1. Generate embedding for the comparison topic
  const topicEmbedding = await embeddingService.generateEmbedding(topic);

  // 2. Search each document for relevant chunks
  const [doc1Results, doc2Results] = await Promise.all([
    vectorStoreService.searchByDocument(topicEmbedding, documentIds[0], topK),
    vectorStoreService.searchByDocument(topicEmbedding, documentIds[1], topK)
  ]);

  if (doc1Results.length === 0 && doc2Results.length === 0) {
    const emptyStructured = {
      similarities: [],
      differences: [],
      uniqueToDoc1: [],
      uniqueToDoc2: [],
      summary: {
        overallAssessment: 'No relevant content found in either document for the given topic.',
        agreementLevel: 'none',
        keyTakeaway: 'Upload documents with relevant content and try again.'
      },
      metadata: {
        doc1ChunksAnalyzed: 0,
        doc2ChunksAnalyzed: 0,
        topic,
        comparisonConfidence: 0
      }
    };

    return {
      comparison: useStructured ? emptyStructured : 'No relevant content found in either document for the given topic.',
      structured: useStructured,
      doc1Sources: [],
      doc2Sources: [],
      topic,
      processingTime: `${((Date.now() - startTime) / 1000).toFixed(2)}s`
    };
  }

  // 3. Generate comparison with LLM (structured or plain text)
  let comparison;
  if (useStructured) {
    comparison = await llmService.generateStructuredComparison(topic, doc1Results, doc2Results);
  } else {
    comparison = await llmService.generateComparison(topic, doc1Results, doc2Results);
  }

  // 4. Build source citations for each document
  const buildSources = (results) =>
    results.map(r => ({
      filename: r.metadata?.filename,
      chunkIndex: r.metadata?.chunkIndex,
      documentId: r.metadata?.documentId,
      similarityScore: parseFloat((r.similarityScore || 0).toFixed(4)),
      preview: r.text?.substring(0, 200) + (r.text?.length > 200 ? '...' : '')
    }));

  const elapsed = ((Date.now() - startTime) / 1000).toFixed(2);
  console.log(`[RAG]   Comparison generated in ${elapsed}s`);

  return {
    comparison,
    structured: useStructured,
    doc1Sources: buildSources(doc1Results),
    doc2Sources: buildSources(doc2Results),
    topic,
    documentsCompared: documentIds,
    processingTime: `${elapsed}s`
  };
}

// ──────────────────────────────────────────
// Document Management
// ──────────────────────────────────────────

/**
 * Delete a document and all its chunks from the vector store.
 */
async function deleteDocument(documentId) {
  await vectorStoreService.deleteDocument(documentId);
  
  // Invalidate query cache since document corpus changed
  cacheService.invalidateDocumentQueries(documentId);
  
  return { documentId, message: 'Document deleted successfully' };
}

/**
 * List all ingested documents.
 */
async function listDocuments() {
  return vectorStoreService.listDocuments();
}

/**
 * Get vector store statistics.
 */
async function getStats() {
  return vectorStoreService.getStats();
}

module.exports = {
  ingestDocument,
  queryDocuments,
  compareDocuments,
  deleteDocument,
  listDocuments,
  getStats
};

File: app/services/rag/rag_service.py
"""
RAG Orchestration Service
Ties together: parse → chunk → embed → store → search → rerank → LLM answer.
"""
import asyncio
import time
import uuid
from typing import Any, Dict, List, Optional

from app.services.embeddings import embedding_service
from app.services.llm import llm_service
from app.services.rag import rerank_service
from app.vectorstore import vector_store_service
from app.utils import cache_service
from app.utils.document_parser import parse_document
from app.utils.chunking_service import (
    chunk_text,
    create_chunk_metadata,
    DEFAULT_CHUNK_SIZE,
    DEFAULT_CHUNK_OVERLAP,
)


# ──────────────────────────────────────────────────────────
# Document Ingestion Pipeline
# ──────────────────────────────────────────────────────────

async def ingest_document(
    filename: str,
    content: bytes,
    chunk_size: Optional[int] = None,
    chunk_overlap: Optional[int] = None,
) -> Dict[str, Any]:
    """
    Full upload pipeline: parse → chunk → embed → store.
    """
    document_id = str(uuid.uuid4())
    start = time.time()

    print(f'[RAG] Ingesting "{filename}" (id: {document_id})')

    # 1. Parse the document into raw text
    raw_text = await parse_document(filename, content)
    print(f"[RAG]   Parsed: {len(raw_text)} characters")

    # 2. Chunk the text
    chunks = chunk_text(raw_text, chunk_size=chunk_size, chunk_overlap=chunk_overlap)
    print(f"[RAG]   Chunked: {len(chunks)} chunks")

    if not chunks:
        raise ValueError("Document produced zero chunks after parsing")

    # 3. Create chunk metadata
    chunk_meta = create_chunk_metadata(document_id, filename, chunks)

    # 4. Generate embeddings for all chunks (batched)
    embeddings = await embedding_service.generate_embeddings_batched(chunks, batch_size=20)
    print(f"[RAG]   Embedded: {len(embeddings)} vectors")

    # 5. Store in ChromaDB
    vector_store_service.store_chunks(chunk_meta, embeddings)

    # 6. Invalidate query cache since corpus changed
    cache_service.invalidate_queries()

    elapsed = f"{time.time() - start:.2f}s"
    print(f"[RAG]   Ingestion complete in {elapsed}")

    return {
        "document_id": document_id,
        "filename": filename,
        "chunk_count": len(chunks),
        "character_count": len(raw_text),
        "chunk_config": {
            "chunk_size": chunk_size or DEFAULT_CHUNK_SIZE,
            "chunk_overlap": chunk_overlap or DEFAULT_CHUNK_OVERLAP,
        },
        "processing_time": elapsed,
        "message": f'Successfully ingested "{filename}" into {len(chunks)} chunks',
    }


# ──────────────────────────────────────────────────────────
# Query Pipeline
# ──────────────────────────────────────────────────────────

async def query_documents(
    query: str,
    top_k: int = 5,
    document_id: Optional[str] = None,
    temperature: Optional[float] = None,
    include_metadata: bool = True,
    verify: bool = False,
    rerank: bool = False,
    skip_cache: bool = False,
) -> Dict[str, Any]:
    """
    Full query pipeline: embed query → search → rerank → LLM answer.
    """
    skip_cache = skip_cache or verify
    start = time.time()

    # Check cache
    if not skip_cache:
        cache_opts = {"top_k": top_k, "document_id": document_id, "rerank": rerank}
        cached = cache_service.get_query_result(query, cache_opts)
        if cached:
            print(f'[RAG] Query served from cache: "{query[:50]}..."')
            cached["processing_time"] = "0.00s (cached)"
            return cached

    print(f'[RAG] Query: "{query[:80]}..." (topK={top_k}, verify={verify}, rerank={rerank})')

    # 1. Embed the query
    query_embedding = await embedding_service.generate_embedding(query)

    # 2. Search for relevant chunks (fetch extra if reranking)
    fetch_k = min(top_k * 3, 20) if rerank else top_k
    if document_id:
        results = vector_store_service.search_by_document(query_embedding, document_id, fetch_k)
    else:
        results = vector_store_service.search_similar(query_embedding, fetch_k)

    if not results:
        return {
            "answer": "No relevant documents found. Please upload documents first.",
            "confidence": {"score": 0, "reason": "No documents found", "level": "none"},
            "sources": [], "query": query, "reranked": False,
            "processing_time": f"{time.time() - start:.2f}s",
        }

    # 3. Optionally re-rank
    if rerank and len(results) > 1:
        print(f"[RAG]   Re-ranking {len(results)} chunks...")
        results = rerank_service.rerank_chunks(query, results, top_n=top_k)
    elif len(results) > top_k:
        results = results[:top_k]

    # 4. Generate answer
    raw_answer = await llm_service.generate_answer(query, results, temperature=temperature)

    # 5. Parse confidence
    answer, confidence = llm_service.parse_confidence(raw_answer)

    # 6. Optionally verify
    verification = None
    if verify:
        print("[RAG]   Running answer verification...")
        verification = await llm_service.verify_answer(answer, results)
        v_status = "PASSED" if verification.get("isVerified") else "FAILED"
        print(f"[RAG]   Verification: {v_status} ({verification.get('overallScore', 0)}/10)")

    # 7. Build source citations
    sources = []
    for r in results:
        meta = r.get("metadata", {})
        src = {
            "filename": meta.get("filename"),
            "chunk_index": meta.get("chunk_index"),
            "document_id": meta.get("document_id"),
            "similarity_score": round(r.get("similarity_score", 0), 4),
        }
        if r.get("rerank_score") is not None:
            src["rerank_score"] = r["rerank_score"]
            src["original_rank"] = r.get("original_rank")
        if include_metadata:
            text = r.get("text", "")
            src["preview"] = text[:200] + ("..." if len(text) > 200 else "")
        sources.append(src)

    elapsed = f"{time.time() - start:.2f}s"
    print(f"[RAG]   Answered in {elapsed} using {len(results)} chunks (confidence: {confidence.get('score')}/10)")

    response: Dict[str, Any] = {
        "answer": answer,
        "confidence": confidence,
        "sources": sources,
        "query": query,
        "top_k": top_k,
        "reranked": rerank,
        "chunks_used": len(results),
        "processing_time": elapsed,
    }
    if verification:
        response["verification"] = verification

    # Cache result
    if not skip_cache:
        cache_opts = {"top_k": top_k, "document_id": document_id, "rerank": rerank}
        cache_service.set_query_result(query, cache_opts, response)

    return response


# ──────────────────────────────────────────────────────────
# Compare Pipeline
# ──────────────────────────────────────────────────────────

async def compare_documents(
    document_ids: List[str],
    topic: str,
    top_k: int = 5,
    structured: bool = False,
) -> Dict[str, Any]:
    """
    Compare two documents on a given topic.
    """
    start = time.time()
    print(f'[RAG] Comparing docs [{", ".join(document_ids)}] on: "{topic[:60]}" (structured={structured})')

    # 1. Embed topic
    topic_embedding = await embedding_service.generate_embedding(topic)

    # 2. Search each document
    doc1_results = vector_store_service.search_by_document(topic_embedding, document_ids[0], top_k)
    doc2_results = vector_store_service.search_by_document(topic_embedding, document_ids[1], top_k)

    if not doc1_results and not doc2_results:
        empty = {
            "similarities": [], "differences": [],
            "uniqueToDoc1": [], "uniqueToDoc2": [],
            "summary": {
                "overallAssessment": "No relevant content found in either document.",
                "agreementLevel": "none",
                "keyTakeaway": "Upload documents with relevant content and try again.",
            },
            "metadata": {"doc1ChunksAnalyzed": 0, "doc2ChunksAnalyzed": 0, "topic": topic},
        }
        return {
            "comparison": empty if structured else "No relevant content found.",
            "structured": structured,
            "doc1_sources": [], "doc2_sources": [],
            "topic": topic, "documents_compared": document_ids,
            "processing_time": f"{time.time() - start:.2f}s",
        }

    # 3. Generate comparison
    if structured:
        comparison = await llm_service.generate_structured_comparison(topic, doc1_results, doc2_results)
    else:
        comparison = await llm_service.generate_comparison(topic, doc1_results, doc2_results)

    # 4. Build citations
    def build_sources(results):
        return [
            {
                "filename": r.get("metadata", {}).get("filename"),
                "chunk_index": r.get("metadata", {}).get("chunk_index"),
                "document_id": r.get("metadata", {}).get("document_id"),
                "similarity_score": round(r.get("similarity_score", 0), 4),
                "preview": r.get("text", "")[:200],
            }
            for r in results
        ]

    elapsed = f"{time.time() - start:.2f}s"
    print(f"[RAG]   Comparison generated in {elapsed}")

    return {
        "comparison": comparison,
        "structured": structured,
        "doc1_sources": build_sources(doc1_results),
        "doc2_sources": build_sources(doc2_results),
        "topic": topic,
        "documents_compared": document_ids,
        "processing_time": elapsed,
    }


# ──────────────────────────────────────────────────────────
# Document Management
# ──────────────────────────────────────────────────────────

def delete_document(document_id: str) -> Dict:
    vector_store_service.delete_document(document_id)
    cache_service.invalidate_document_queries(document_id)
    return {"document_id": document_id, "message": "Document deleted successfully"}


def list_documents() -> List[Dict]:
    return vector_store_service.list_documents()


def get_stats() -> Dict:
    return vector_store_service.get_stats()
